src/main/scala/chisel3/util/LTL.scala

```diff
@@ -0,0 +1,392 @@
+  * Logic (LTL) intrinsics operate on `Bool` values instead of a more desirable
+  * `Sequence` type. To prevent abuse, the Chisel API here wraps these `Bool`s
+  * in an opaque type over which we can provide a safe interface that cannot be
+  * used to construct invalid IR.
+  */
+private case class OpaqueSequence(private[ltl] val inner: Bool) extends Sequence
+
+/** An opaque property returned by an intrinsic.
+  *
+  * See `OpaqueSequence` for details.
+  */
+private case class OpaqueProperty(private[ltl] val inner: Bool) extends Property
+
+/** A single item that may be used in the `Sequence(...)` convenience
+  * constructor for sequences. These atoms may either be `Sequence`s themselves,
+  * like `a` or `a and b`, or a `DelayAtom`, like `###`. Together they enable a
+  * shorthand notation for sequences like:
+  *
+  * `Sequence(a, ###, b, ###(2), c, ###(3, 9), d, ###(4, None))`.
+  */
+sealed trait SequenceAtom
+
+/** A delay atom for the `Sequence(...)` convenience constructor. * */
+private case class DelayAtom(val min: Int, val max: Option[Int]) extends SequenceAtom
+
+/** The delay atoms available to users. Can be interleaved with actual sequences
+  * in `Sequence(...)`. See `SequenceAtom` for details.
+  */
+object ### extends SequenceAtom {
+  def apply(delay: Int): SequenceAtom = DelayAtom(delay, Some(delay))
+  def apply(min:   Int, max: Int): SequenceAtom = DelayAtom(min, Some(max))
+  def apply(min:   Int, max: Option[Int]): SequenceAtom = DelayAtom(min, max)
+}
+
+/** A Linear Temporal Logic (LTL) sequence. */
+trait Sequence extends Property {
+
+  /** See `Sequence.delay`. */
+  def delay(delay: Int = 1): Sequence = Sequence.delay(this, delay)
+
+  /** See `Sequence.delayRange`. */
+  def delayRange(min: Int, max: Int): Sequence = Sequence.delayRange(this, min, max)
+
+  /** See `Sequence.delayAtLeast`. */
+  def delayAtLeast(delay: Int): Sequence = Sequence.delayAtLeast(this, delay)
+
+  /** See `Sequence.concat`. */
+  def concat(other: Sequence): Sequence = Sequence.concat(this, other)
+
+  /** See `Sequence.and`. */
+  def and(other: Sequence): Sequence = Sequence.and(this, other)
+
+  /** See `Sequence.or`. */
+  def or(other: Sequence): Sequence = Sequence.or(this, other)
+
+  /** See `Sequence.clock`. */
+  override def clock(clock: Clock): Sequence = Sequence.clock(this, clock)
+
+  /** See `Property.implication`. */
+  def implication(prop: Property): Property = Property.implication(this, prop)
+
+  /** See `Property.implication`. */
+  def implicationNonOverlapping(prop: Property): Property = Property.implicationNonOverlapping(this, prop)
+
+  // Convenience operators.
+
+  /** Equivalent to `|->` in SVA. */
+  def |->(prop: Property): Property = this.implication(prop)
+
+  /** Equivalent to `|=>` in SVA. */
+  def |=>(prop: Property): Property = this.implicationNonOverlapping(prop)
+
+  /** Equivalent to `a ## b` in SVA. */
+  def ###(other: Sequence): Sequence = this.concat(other.delay())
```

Comment:
With my above comment, this definition also means that `###` is being used in two different ways in the code.

Reply:
Conceptually they are both infix operators. The `###` here is a true infix operator in the Scala sense, but it has the delay hardwired to 1, since you can't have additional parameters on the infix operator. For the more complicated case you need `###` in `Sequence(...)`, which technically isn't an infix operator from Scala's point of view, but just an individual object; however the way it's used inside of `Sequence(...)` is still with infix notation, just with `,` sprinkled in.